text/0039-source-removal-buildpack.md

```diff
@@ -0,0 +1,61 @@
+# Source Removal Buildpack
+
+## Summary
+
+Adopt [ForestEckhardt's Source Removal buildpack](https://github.com/ForestEckhardt/source-removal)
+into the Paketo Project to allow for uniform source removal logic across all
+buildpacks that require such functionality.
+
+## Motivation
+
+There are several buildpacks in the Paketo Buildpacks project that take their
+given source code and transform that into a new artifact such as a binary or
+other forms of compiled/transpiled code. It is often desirable to remove any
+unnecessary source code from the final image as it causes image bloat and poses
+a potential security risk. In order to currently accomplish this, any buildpack
+that wants to remove source code must have logic to remove it from the
+`workspace` so that it does not appear in the final image, however this task is
+complicated because often times there are static files that are required for
+the app to run that should no be deleted from the `workspace`. This leads to
+multiple implementations of non-trivial file removal logic across different
+buildpacks.
+
+This could all be simplified by adding a buildpack that performs all source
+removal logic to the end of any buildpack order that would require things to be
+removed from the `workspace` which is exactly what ForestEckhardt's Source
+Removal buildpack does. This buildpacks allows this shared logic to be
+centralized into one place and then reused across the entire buildpack
+ecosystem and as needed by particular users.
+
+## Rationale and Alternatives
+
+- Continue to implement source removal logic in all buildpacks that would
+  benefit from such functionality.
+
+## Implementation
+
+Currently ForestEckhardt's Source Removal buildpack always passes detection and
+has three actions.
+
+1. If no configuration is set (i.e. `BP_INCLUDE_FILES` and `BP_EXCLUDE_FILES`
+   are not set), then all the content of the `workspace` will be deleted.
+2. If `BP_INCLUDE_FILES` is set, then all files that **do not** match the
+   pattern globs specified in the environment variables will be deleted.
```

Comment:
Possibly a note that is more for the buildpack docs, but I think we should define and provide some examples of what is meant by "globs". Perhaps some examples of how you can include/exclude various files using globs.

An alternative would be to link to a glob library or shell docs that explain how this would work. If we can say, we use library X or we implement things like Shell Y, then users will have a more concrete idea of what is possible.

Reply:
I am more than happy to reveal that information if it should be necessary. Currently I am just using [`filepath.Match`](https://pkg.go.dev/path/filepath#Match) which is part of the Go standard library.